### PR TITLE
OpenCL: Fix duplication of layers in VRAM and RAM, add GPU mul kernel

### PR DESCRIPTION
**Pull request by 0cc4m**
Port further improvements from the CUDA version to OpenCL, specifically:

* No more duplication of layers, they will now be solely in VRAM if offloaded to a GPU
* The norm calculation is now also done on GPU for that reason

### CODE REVIEW

**Review by ggerganov — APPROVED** [review]
Haven't tested, the `ggml` part is OK

**Comment by LostRuins** [comment]
Hi, I can confirm this works when used directly on @0cc4m 's branch. 

However, it fails when merged into master due to malloc changes from @howard0su done in https://github.com/ggerganov/llama.cpp/pull/1612 resulting in the error `ggml_opencl: clSetKernelArg(*to_fp32_cl, 0, sizeof(cl_mem), &d_Q) error -38 at ggml-opencl.cpp:1029`

Performance wise, there is still a minor (<5%) speed regression, but the benefit of saving of saving a large chunk of RAM may be worth the tradeoff. (Apparently CUDA's implementation has similar regressions too.)

**Comment by 0cc4m** [comment]
I didn't notice the changes in #1612, should be fixed now.

**Comment by SlyEcho** [comment]
Tested working

**Comment by LostRuins** [comment]
It's working now. @SlyEcho do you observe and speed changes?

**Comment by SlyEcho** [comment]
I do not see a performance regression.

However I can get a massive boost by changing `CL_DMMV_BLOCK_SIZE` to 64.

Generation goes from 75ms/t to 58ms/t

Thisi is with both versions.

**Comment by SlyEcho** [comment]
I think it depends on the GPU. I can even use 128, but 256 breaks it.

**Comment by 0cc4m** [comment]
@SlyEcho Did you use a GCN card? I see a very small reduction in speed with 64, probably because of this:
> GCN in all of its iterations was 64 threads wide, meaning 64 threads were bundled together into a single wavefront for execution. RDNA drops this to a native 32 threads wide.

File: llama.cpp [hunk 1]
```diff
@@ -1093,15 +1095,23 @@ static void llama_model_load_internal(
...
-#elif !defined(GGML_USE_CLBLAST)
+#elif defined(GGML_USE_CLBLAST)
+        const int n_gpu = std::min(n_gpu_layers, int(hparams.n_layer));
+
+        fprintf(stderr, "%s: [opencl] offloading %d layers to GPU\n", __func__, n_gpu);
+        if (n_gpu_layers > (int) hparams.n_layer) {
+            fprintf(stderr, "%s: [opencl] offloading output layer to GPU\n", __func__);
+        }
+        fprintf(stderr, "%s: [opencl] total VRAM used: %zu MB\n", __func__, vram_total / 1024 / 1024);
+#else
```
Comment (howard0su) [inline llama.cpp:1114]: Shall we meger two branch of #if? they are looks same except the prefix which didn't give that much information. we can add one line ahead says we are using CUDA or OpenCL to do offloading.

Comment (0cc4m) [reply llama.cpp:1114]: Probably a good idea, yeah.

File: ggml-opencl.cpp [hunk 2]
```diff
@@ -862,42 +985,46 @@ static void ggml_cl_mul_mat_q_f32(const ggml_tensor * src0, const ggml_tensor *
 
     for (int64_t i03 = 0; i03 < ne03; i03++) {
         for (int64_t i02 = 0; i02 < ne02; i02++) {
-            cl_event ev_sgemm;
+            size_t ev_idx = 0;
+            std::vector<cl_event> events;
```
Comment (howard0su) [inline ggml-opencl.cpp:989]: suggest move this out of loop and give a reasonable initial size. You can use clear in the inner-loop

**Comment by 0cc4m** [comment]
I've improved it according to the reviews, anything else that should be fixed? Otherwise I think it's good enough now.

**Review by howard0su — APPROVED** [review]

**Review by SlyEcho — APPROVED** [review]

**Comment by LostRuins** [comment]
Bump, I think this should be merged, works well for me.

**Pull request merged**